### PR TITLE
Added Incremental Merkle Tree implementation

### PR DESCRIPTION
**Pull request by KyrylR**
- [ ] Since this PR suggests a **bug fix**, the tests have been added and the coverage is 100%.
- [x] Since this PR introduces a **new feature**, the update has been discussed in an Issue or with the team.
- [ ] This PR is just a **minor change**, like a typo fix.

---

### CODE REVIEW

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+ * uintTree.addUint(1234);
+ *
+ * uintTree.getRoot();
+ *
+ * uintTree.getTreeHeight();
+ */
+library IncrementalMerkleTree {
+    /**
+     ************************
+     *      UintVector      *
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:37]:
```suggestion
     *      UintIMT      *
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+     *      InnerVector     *
+     ************************
+     */
+
+    struct IMT {
+        bytes32[] branches;
+        uint256 depositCount;
+    }
+
+    bytes32 public constant ZERO_HASH = keccak256(abi.encode(0));
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:132]: This should not be public

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+     ************************
+     *     AddressVector    *
+     ************************
+     */
+
+    struct AddressIMT {
+        IMT _tree;
+    }
+
+    function addAddress(AddressIMT storage tree, address element_) internal {
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:109]:
```suggestion
    function add(AddressIMT storage tree, address element_) internal {
```

Comment (Arvolear) [reply contracts/libs/data-structures/IncrementalMerkleTree.sol:109]: And others. Function overloading will do the job

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+     */
+
+    struct IMT {
+        bytes32[] branches;
+        uint256 depositCount;
+    }
+
+    bytes32 public constant ZERO_HASH = keccak256(abi.encode(0));
+
+    function _addBytes32(IMT storage tree, bytes32 element_) private {
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:134]:
```suggestion
    function _add(IMT storage tree, bytes32 element_) private {
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+            assembly {
+                mstore(0, branch_)
+                mstore(32, resultValue_)
+
+                resultValue_ := keccak256(0, 64)
+            }
+
+            size_ >>= 1;
+
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:161]:
```suggestion
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+            return ZERO_HASH;
+        }
+
+        bytes32[] memory zeroHashes_ = _getZeroHashes(treeHeight_);
+
+        bytes32 root_ = ZERO_HASH;
+
+        uint256 size_ = tree.depositCount;
+
+        uint256 height_;
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:185]:
```suggestion
        bytes32[] memory zeroHashes_ = _getZeroHashes(treeHeight_);
        bytes32 root_ = ZERO_HASH;
        uint256 size_ = tree.depositCount;
        uint256 height_;
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+                assembly {
+                    mstore(0, root_)
+                    mstore(32, zeroHash_)
+
+                    root_ := keccak256(0, 64)
+                }
+            }
+
+            size_ >>= 1;
+
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:209]:
```suggestion
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+            height_++;
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:210]:
```suggestion
            ++height_;
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+            assembly {
+                mstore(0, branch_)
+                mstore(32, resultValue_)
+
+                resultValue_ := keccak256(0, 64)
+            }
+
+            size_ >>= 1;
+
+            index_++;
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:162]:
```suggestion
            ++index_;
```

File: contracts/mock/libs/data-structures/IncrementalMerkleTreeMock.sol [hunk 10]
```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import {IncrementalMerkleTree} from "../../../libs/data-structures/IncrementalMerkleTree.sol";
+
+contract IncrementalMerkleTreeMock {
+    using IncrementalMerkleTree for IncrementalMerkleTree.UintIMT;
+    using IncrementalMerkleTree for IncrementalMerkleTree.Bytes32IMT;
+    using IncrementalMerkleTree for IncrementalMerkleTree.AddressIMT;
```
Comment (Arvolear) [inline contracts/mock/libs/data-structures/IncrementalMerkleTreeMock.sol:9]:
```suggestion
    using IncrementalMerkleTree for *;
```

File: test/helpers/merkleTreeHelper.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
+const { web3 } = require("hardhat");
+
```
Comment (Arvolear) [inline test/helpers/merkleTreeHelper.js:2]:
```suggestion
```

I think web3 import can be ignored

File: test/helpers/merkleTreeHelper.js
Comment (Arvolear) [inline test/helpers/merkleTreeHelper.js:1]: Please rename to `merkle-tree-helper`

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+
+    /**
+     ************************
+     *      InnerVector     *
+     ************************
+     */
+
+    struct IMT {
+        bytes32[] branches;
+        uint256 depositCount;
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:129]:
```suggestion
        uint256 leavesCount;
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        return _getTreeHeight(tree._tree);
+    }
+
+    /**
+     ************************
+     *      InnerVector     *
+     ************************
+     */
+
+    struct IMT {
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:127]: Please add function to get the `leavesCount`. Maybe the `length()`?

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+    function addAddress(AddressIMT storage tree, address element_) internal {
+        _addBytes32(tree._tree, bytes32(uint256(uint160(element_))));
+    }
+
+    function getRoot(AddressIMT storage tree) internal view returns (bytes32) {
+        return _getRoot(tree._tree);
+    }
+
+    function getTreeHeight(AddressIMT storage tree) internal view returns (uint256) {
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:117]:
```suggestion
    function height(AddressIMT storage tree) internal view returns (uint256) {
```

File: README.md [hunk 15]
```diff
@@ -14,6 +14,7 @@ The library consists of modules and utilities that are built with a help of [Ope
 - Enhanced and simplified [**Diamond**](https://eips.ethereum.org/EIPS/eip-2535) pattern
 - Heap based priority queue library
 - Memory data structures (Vector)
+- Enhanced Incremental Merkle Tree data structure from  [**Deposit Contract Verification**](https://github.com/runtimeverification/deposit-contract-verification/blob/master/deposit-contract-verification.pdf).
```
Comment (Arvolear) [inline README.md:17]:
```suggestion
- Enhanced [**Incremental Merkle Tree**](https://github.com/runtimeverification/deposit-contract-verification/blob/master/deposit-contract-verification.pdf) data structure
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 16]
```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+/**
+ * @title Incremental Merkle Tree module
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:5]:
```suggestion
 * @notice Incremental Merkle Tree module
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+ *
+ * This implementation is a modification of the Incremental Merkle Tree data structure mentioned
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:7]:
```suggestion
 * This implementation is a modification of the Incremental Merkle Tree data structure described
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+ * in [Deposit Contract Verification](https://github.com/runtimeverification/deposit-contract-verification/blob/master/deposit-contract-verification.pdf).
+ *
+ * This implementation aims to optimise and improve the original data structure.
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:10]:
```suggestion
 * This implementation aims to optimize and improve the original data structure.
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+ *
+ * The main differences are:
+ * - No explicit constructor; the tree is initialised when the first element is added.
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:13]:
```suggestion
 * - No explicit constructor; the tree is initialised when the first element is added
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+ * - Growth is not constrained; the height of the tree grows as elements are added
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:14]:
```suggestion
 * - Growth is not constrained; the height of the tree automatically increases as elements are added
```

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+ *
+ * Zero hashes are computed each time the getRoot function is called.
+ *
+ * You can find the gas analysis [here](https://colab.research.google.com/drive/1I-nGhb_LdcWW4pvQtE2F9JVDv371rO5f?usp=sharing).
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:18]: Let's not include personal external links for now. We will have a DLSL site in the future. You can include the gas usage right here

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+ *
+ * IncrementalMerkleTree.UintIMT internal uintTree;
+ *
+ * ................................................
+ *
+ * uintTree.addUint(1234);
+ *
+ * uintTree.getRoot();
+ *
+ * uintTree.getTreeHeight();
```
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:32]: This has to be updated according to the comments

File: contracts/libs/data-structures/IncrementalMerkleTree.sol
Comment (Arvolear) [inline contracts/libs/data-structures/IncrementalMerkleTree.sol:1]: A minor comment. Right now we are writing comments starting with 2 whitespaces after the *

File: test/libs/data-structures/IncrementalMerkleTree.js [hunk 23]
```diff
@@ -0,0 +1,145 @@
+const { assert } = require("chai");
+
```
Comment (Arvolear) [inline test/libs/data-structures/IncrementalMerkleTree.js:2]:
```suggestion
```

File: contracts/mock/libs/data-structures/IncrementalMerkleTreeMock.sol [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+
+    IncrementalMerkleTree.UintIMT internal uintTree;
```
Comment (Arvolear) [inline contracts/mock/libs/data-structures/IncrementalMerkleTreeMock.sol:11]: Internal naming?

File: test/libs/data-structures/IncrementalMerkleTree.js (hunk 23 above)
Comment (KyrylR) [reply test/libs/data-structures/IncrementalMerkleTree.js:2]: I think it's better to visually distinguish between different import groups (from node, local files, etc.) 
What do you think?

Comment (Arvolear) [reply test/libs/data-structures/IncrementalMerkleTree.js:2]: Maybe, but let's stick to the already existing tests

**Pull request merged**